src/lib/reasoners/theory.ml

```diff
@@ -338,110 +340,244 @@ module Main_Default : S = struct
     cs_pending_facts : (E.t * Ex.t * int * int) list list;
     terms : Expr.Set.t;
     gamma : CC_X.t;
     gamma_finite : CC_X.t;
-    choices : choice list
+    choices : choice list;
+    objectives : Th_util.optimized_split Util.MI.t;
   }
 
-  let add_explanations_to_splits l =
-    List.map
-      (fun (c, is_cs, size) ->
-         Steps.incr_cs_steps();
-         let exp = Ex.fresh_exp () in
-         let ex_c_exp =
-           if is_cs then Ex.add_fresh exp Ex.empty else Ex.empty
-         in
-         (* A new explanation in order to track the choice *)
-         (c, size, CPos exp, ex_c_exp)) l
+  let add_explanations_to_split (c, is_cs, size) =
+    Steps.incr_cs_steps ();
+    let exp = Ex.fresh_exp () in
+    let ex_c_exp =
+      if is_cs then Ex.add_fresh exp Ex.empty else Ex.empty
+    in
+    (* A new explanation in order to track the choice *)
+    (c, size, CPos exp, ex_c_exp, None)
+
+  let register_optimized_split objectives u =
+    try
+      let x = Util.MI.find u.Th_util.order objectives in
+      assert (E.equal x.Th_util.e u.Th_util.e);
+      (* and its Value ... *)
+      Util.MI.add u.Th_util.order u objectives
+    with Not_found ->
+      assert false
+
+  exception Found of Th_util.optimized_split
+
+  (* TODO: this function could be optimized if "objectives" structure
+     is coded differently *)
+  let next_optimization ~for_model env =
+    try
+      Util.MI.iter (fun _ x ->
+          match x.Th_util.value with
+          | Value _ ->
+            (* This split is already optimized. *)
+            ()
+          | Pinfinity | Minfinity | Limit _ ->
+            (* We should block case-split at infinite values.
+                  Otherwise, we may have soundness issues. We
+                  may think an objective is unbounded, but some late
+                  constraints may make it bounded.
+
+                  An alternative is to only allow further splits when we
+                  know that no extra assumptions will be propagated to
+                  the env. Hence the test 'if for_model' *)
+            if for_model then ()
+            else
+              raise (Found { x with Th_util.value = Unknown })
+
+          | Unknown ->
+            raise (Found { x with Th_util.value = Unknown })
+
+        ) env.objectives;
+      None
+    with
+    | Found x -> Some x
+
+  (* TODO: this function could be optimized if "objectives" structure
+     is coded differently *)
+  let partial_objectives_reset objectives order =
+    match order with
+    | None -> objectives
+    | Some opt_ord ->
+      Util.MI.fold
+        (fun ord v acc ->
+           if ord < opt_ord then
+             (*don't change older optims that are still valid splits*)
+             acc
+           else
+             match v.Th_util.value with
+             | Th_util.Unknown -> acc (* not optimized yet *)
+             | Value _ | Limit _ ->
+               Util.MI.add ord {v with value = Unknown} acc
```

Comment:
I don't think the `Limit` case can happen here? Should be with `Pinfinity` and `Minfinity` (we block optims under all of these -- see `next_optimization`).